### PR TITLE
Fix matrix_box cache invalidation on nested records

### LINKED ISSUE #1907: Name deprecations do not unbold names in cached matrix boxes on observation indexes
Sometimes light (correct), but sometimes bold (wrong).
Examples: 
- https://mushroomobserver.org/observations?pattern=name%3A%22Trichia+decipiens%22
- goto https://mushroomobserver.org/names/21206 
In the "About this taxon" panel, click "this name"

### PR DESCRIPTION
**Pull request by nimmolo**
This PR 
- Fixes #1907
- Restores caching for matrix_boxes on obs and rss_logs indexes
- Modifies an existing `before_update` callback for Names and Locations, to check whether anything relevant to an Observation matrix_box has changed. If so, it touches associated Observations to invalidate HTML caches. It does the same if a User `login` gets changed.  
___
Took awhile to track down, but turns out we already have `before_update` callbacks for Name and Location that update cached columns in Observation for `text_name`, `where`, and others when the associations change. 

By altering these callbacks to also touch `updated_at`, we can invalidate the HTML caches of observations without running any additional query. 

To cover the very rare case where a User changes their login, I've added a callback in User too.

### CODE REVIEW

**Review by mo-nathan — APPROVED** [review]
This looks good!  I even follow most of it.  I can't manually test it since I still don't have caching working locally.

File: app/models/name/notify.rb [hunk 1]
```diff
@@ -1,9 +1,19 @@
...
+      sender_email: user.email,
+      subject: "#{user.login} created #{real_text_name}",
+      content: "#{MO.http_domain}/names/#{id}"
+    )
+  end
+
   # This is called after saving potential changes to a Name.  It will determine
   # if the changes are important enough to notify the authors, and do so.
-  def notify_users
+  def notify_users # rubocop:disable Metrics/MethodLength
```
[the remark this answers is not in the document]
Comment (nimmolo) [reply app/models/name/notify.rb]: Fixed

**Comment by nimmolo** [comment]
@JoeCohen I will hold this until you confirm it's working on your dev environment, since you have the cache too.

**Review by JoeCohen — APPROVED** [review]
@nimmolo: 
I don't know how to reproduce the bug of #1907 locally. (Ideally, I'd be able to see that the bug appears in main, but is fixed in this branch.)
In my environment, I'm not seeing any issues with this branch.
So I think you should feel free to merge it.
But if you have an easy way for me to reproduce the bug, I'd be happy to have another look.

**Comment by nimmolo** [comment]
@JoeCohen Thanks for checking it out. I turned off the cache for matrix boxes the other day in #1913, that's what made the issue disappear. I think your dev environment was still messed up, and you couldn't check out that PR?

So anyway - there's no way to reproduce the bug easily. Will merge. 
This will be an interesting off/on test case for the matrix box caches, since they've been off for a few days.

**Comment by mo-nathan** [comment]
Verified with Trichia decipiens which was deprecated but there were still cached examples that were bold.  Switched to accepted and then deprecated it again and the cache got cleared.  Thanks for the work on this!

**Pull request merged**